README.md
# Cookie Banner [![Build Status](https://img.shields.io/travis/dobarkod/cookie-banner.svg?style=flat-square)](https://travis-ci.org/dobarkod/cookie-banner)

Cookie Banner is a super-easy way to ensure you're complying with the EU
cookie law. Just reference the `cookiebanner.min.js` script from your page
and you're done.

Cookie Banner script is very lightweight and depends on no JavaScript
libraries, css files or images.

## Demo

To see the script in action [download the files](https://github.com/dobarkod/cookie-banner/archive/master.zip) and open/serve [tests/demo.html](tests/demo.html) and/or [tests/tests.html](tests/tests.html).

## Quickstart

Add a single line to your web page, just before the closing `</body>` tag:

```html
<script type="text/javascript" id="cookiebanner"
  src="https://cdn.jsdelivr.net/gh/dobarkod/cookie-banner@1.2.2/dist/cookiebanner.min.js"></script>
```

or via [cdnjs](https://cdnjs.com/):
```html
<script type="text/javascript" id="cookiebanner"
  src="https://cdnjs.cloudflare.com/ajax/libs/cookie-banner/1.2.2/cookiebanner.min.js"></script>
```

This will display a black-and-white floating banner at the bottom of your
web page, informing the user that the site is using cookies, and giving them
the link to a page with more information.

When the user clicks the "close" button, the banner will set a cookie
(oh, the irony!) remembering that the banner was acknowledged, so the same
user will not be bothered again.

## Options

If you don't like the defaults, you can modify the banner content and
style. The following options are settable through a `data-` property on the
`script` tag:

* `height` - Banner/notice height (default: `32px`)
* `fg` - Banner/notice text color (default: `#ddd`)
* `bg` - Banner/notice background color (default: `#000`)
* `link` - Link text color (default: `#aaa`)
* `position` - Banner/notice position, `top` or `bottom` (default: `bottom`)
* `padding` - Set custom padding (default: `5px 16px`)
* `message` - Notice message text
* `linkmsg` - Link text content (default: `Learn more`)
* `close-text` - Text/symbol for the `.cookiebanner-close` element (default: `&#10006;`)
* `close-style` - CSS style for `.cookiebanner-close` element (default: `float:right;padding-left:5px;`)
* `close-precedes` - Controls whether `.cookiebanner-close` element precedes or follows the `<span>` containing the message (default: `true`)
* `font-size` - Text/font size for the .cookiebanner (container) element (default: `14px`)
* `font-family` - Font family for the .cookiebanner (container) element (default: `arial, sans-serif`)
* `text-align` - Text align/position for the .cookiebanner (container) element (default: `center`)
* `effect` - Effect used when inserting the notice, currently only `fade` is supported (triggered when cookiebanner is displayed and when it is closed) (default: `null`)
* `fade-out-duration-ms` - Duration in milliseconds of fade-out effect (only if `fade` effect is specified, default: `2000`)
* `cookie` - Name for the cookie that stores acceptance info (default: `cookiebanner-accepted`)
* `expires` - Cookie expiry date/time (defaults to `Infinity` aka `Fri, 31 Dec 9999 23:59:59 GMT`). There's basic support for specifying a callback function (more flexibility, must return one of `Number`, `String`, or `Date` -- see `Cookies.set()`). You can also just specify a valid UTC string. When specified as a "numeric string", it is treated as number of seconds
from now. So, `expires: '90'` or `data-expires="90"` would expire the cookie in 90
seconds.
* `cookie-path` - Path to set for the cookie
* `cookie-domain` - Set a custom cookie domain (default: `null`)
* `cookie-secure` - Set (`true`/`false`) secure cookie for HTTPS (default: `false`)
* `moreinfo` - Link where the visitor can read more about cookies (default: [http://aboutcookies.org](http://aboutcookies.org))
* `moreinfo-target` - Target for `moreinfo` link (default: `_blank`)
* `moreinfo-decoration` - Text decoration for `moreinfo` link (default: `none`)
* `moreinfo-font-weight` - Font weight for `moreinfo` link (default: `normal`)
* `moreinfo-font-size` - Font size (i.e 12px) for `moreinfo` link (default: `null`)
* `moreinfo-rel` - Rel attribute value for `moreinfo` link (default: `noopener noreferrer`). Set to `false` or `` if you wish to disable it.
* `moreinfo-class` - CSS class for `moreinfo` link (default: `null`). If present, it will be used to detect the `moreinfo` link (instead of getting the first `a` element within `message`). Use when your `message` has more than one link and you do not want the first one to have `moreinfo` link set.
* `mask` - Controls whether a mask is created over the viewport (default: `false`). Clicking anywhere on the mask is considered as acceptance.
* `mask-opacity` - Opacity used for the window `mask` (default: `0.5`)
* `mask-background` - CSS background style applied to the `mask` <div> (default: `#000`)
* `zindex` - Z-index set on the notice (default: `255`). If `mask` is used, the notice <div>'s z-index is automatically incremented by 1 so it appears above the mask.
* `accept-on-scroll` - When `true`, agrees and closes the notice when window is scrolled. (default: `false`)
* `accept-on-click` - When `true`, agrees and closes the notice when clicking anywhere on the page. (default: `false`)
* `accept-on-first-visit` - When `true`, agrees automatically (which stops showing the notice for subsequent requests), but the notice is not automatically closed. (default: `false`)
* `accept-on-timeout` - Automatically agrees and closes the notice after specified number of milliseconds. (default: `null`)
* `on-inserted` - A function which gets executed after the banner is inserted in the DOM (default: `null`)
* `on-closed` - A function which gets executed after the banner is closed (default: `null`)
* `delay-before-close` - Wait N milliseconds before closing the notice. (default: `null`)
* `debug` - When `true`, closes the banner without setting the cookie (default: `false`)

Here's an example:

```html
<script type="text/javascript" id="cookiebanner"
    src="https://cdn.jsdelivr.net/gh/dobarkod/cookie-banner@1.2.2/dist/cookiebanner.min.js"
    data-height="20px" data-position="top"
    data-message="We use cookies to improve your browsing experience.">
</script>
```

## Self-hosting

If you don't want to depend on `cdnjs` or `jsdelivr`, copy
the `cookiebanner.min.js` file directly to your site and serve it yourself.

If you want to compile cookiebanner with the rest of your application's javascript,
you can call it with options as follows:

```html
<script type="text/javascript">
      var options = { message: "We use cookies to enhance your experience.", moreinfo: "/about/cookies" };
      var cb = new Cookiebanner(options); cb.run();
</script>
```

## Internals and more options

If the banner needs to be shown, the script will create the following DOM
subtree by default and add it just before the closing `</body>` tag:

```html
<div class="cookiebanner">
    <div class="cookiebanner-close" style="float: right; padding-left:5px;">&#10006;</div>
    <span>Message. <a rel="noopener noreferrer" href=".." target="_blank">Learn more</a></span>
</div>
```

If the `close-precedes` option is sto to `false` (added in 1.2.1), the DOM looks
slightly different:
```html
<div class="cookiebanner">
    <span>Message. <a rel="noopener noreferrer" href=".." target="_blank">Learn more</a></span>
    <div class="cookiebanner-close" style="float: right; padding-left:5px;">&#10006;</div>
</div>
```

You can use CSS with `div.cookiebanner > span` and `div.cookiebanner > a` to
further modify the banner appearance.

You can also try customizing the close button via the `.cookiebanner-close` CSS class.
Keep in mind that you might have to override and/or reset certain properties by using `!important` CSS rules.


## Event-Handlers (Callbacks)

There are two ways you can utilize event handlers like i.e. `on-inserted`.

Within JS (recommended):

```html
<script type="text/javascript">
      var options = { onInserted: function (instance) { console.log('Hey, I got inserted!') } };
      var cb = new Cookiebanner(options); cb.run();
</script>
```

As a `data`-attribute:

```html
<script type="text/javascript" id="cookiebanner"
    src="https://cdn.jsdelivr.net/gh/dobarkod/cookie-banner/dist/cookiebanner.min.js"
    data-on-inserted="function (instance) { console.log('Hey, I got inserted!'); }">
</script>
```


## Hacking

Get the newest and the freshest from GitHub:
```sh
git clone https://github.com/dobarkod/cookie-banner.git
```

If you've modified the code, it's recommended you run it through linter
to catch potential errors, and minifier to minimize its footprint.

We're old-school here so we just use Makefile for the tasks:

```sh
make lint  # run jshint to check the code
make test  # run the tests
make # minify it
```

You'll need `jshint`, `eslint` and `uglifyjs` (v2) tools installed for this, and also
`phantomjs` if you want to run the automated tests.

Pull requests are welcome! In order to get your pull-request accepted,
please follow these simple rules:

* all code submissions must pass cleanly (no errors) with `make test`
* there should be no external JavaScript, CSS, image files or any other
  dependencies
* the code should work with no errors or warnings on recent Chrome, Firefox,
  Safari, Internet Explorer browsers and iOS and Android platforms
* if you want to add significant features, file an issue about it first so
  we can discuss whether the addition makes sense for the project

## [License (MIT)](LICENSE.md)


CHANGELOG.md
# Change Log

## [Unreleased]
### Added
- New option: `delay-before-close` (default: `null`), allows delaying the notice
closing by the specified number of milliseconds
- New option: `fade-out-duration-ms` (default: `2000`), allows specifying the
fade out duration (in milliseconds)
- `expires` option now handles simple numeric strings as "expires in X number of
seconds from now"
- New options/callbacks: `on-inserted`, `on-closed` (default: `null`) -- see
[README](https://github.com/dobarkod/cookie-banner/blob/master/README.md) for
more details
- New option: `padding` (default: `5px 16px`), allows setting the padding for
cases when one has no control over the site's CSS
- New option: `moreinfo-class` (default: `null`), allows specifying which link
will get the `moreinfo` link (for cases when your `message` has more than one
link, and you do not want the first one to have `moreinfo` link set on it)

### Changed
- `data-effect="fade"` or `effect: 'fade'` now applies when the notice is closed
too (meaning the notice now fades out when closing, if/when it was faded in)

### Fixed
- Empty `linkmsg` option does not throw an error now

## [1.2.2](https://github.com/dobarkod/cookie-banner/compare/1.2.1...1.2.2) - 2018-04-19
### Fixed
- Debug option works now (closes banner without setting the cookie)

## [1.2.1](https://github.com/dobarkod/cookie-banner/compare/1.2.0...1.2.1) - 2017-10-26
### Added
- New option: `close-precedes` (defaults to `true`, keeping backwards compatibility) -- allows controlling whether `.cookiebanner-close` element follows or precedes the `<span>` containing the message

### Fixed
- Tightened up some loose ends in tests
- Added some more eslint rules
- Upgraded eslint dev dependency to ~4.8.0

## [1.2.0](https://github.com/dobarkod/cookie-banner/compare/1.1.1...1.2.0) - 2017-08-01
### Added
- New option: `moreinfo-rel` (controls rel attribute value for `moreinfo` link, defaults to `noopener noreferrer`)

## [1.1.1](https://github.com/dobarkod/cookie-banner/compare/1.1.0...1.1.1) - 2017-03-30
### Fixed
 - Nothing really, just bumping version for cdnjs to pick up latest version

## [1.1.0](https://github.com/dobarkod/cookie-banner/compare/1.0.0...1.1.0) - 2017-03-30
### Added
 - New option: `moreinfo-font-size` (sets font-size for `moreinfo` link when the site's CSS can't be changed)

## [1.0.0](https://github.com/dobarkod/cookie-banner/compare/761222ef3c62efa93c5660a2e1fea52f5e4e2176...1.0.0) - 2017-01-13
### Added
- Now hosted on [CDNJS](https://cdnjs.com/libraries/cookie-banner).
- New option: dismiss the notice by clicking anywhere on the page.
- New option: dismiss the notice as soon as the window is scrolled.
- New option: agree automatically on first visit and show the notice only on first page view.
- New option: agree and close the notice automatically after a certain amount of time passes.
- New option: `close-style`.
- New option: `cookie-secure`.
- New option: `moreinfo-decoration`.
- New option: `moreinfo-font-weight`.
- New option: `moreinfo-target`.
- Minified version included in the `dist/` directory and it contains a short license header.
- This CHANGELOG :)

### Fixed
- Misc. Readme/Documentation changes.
- Fix warnings/notices when running tests on newer phantomjs versions (2.1.x).


tests/chrome-extension/bookmarklet.js
var add_script=function(e,t){var n=document.createElement("script");n.type="text/javascript";n.id="cookiebanner";n.src=e;if(typeof n!=="undefined"){if(t){n.onloadDone=false;n.onload=function(){n.onloadDone=true;t()};n.onReadystatechange=function(){if(n.readyState==="loaded"&&!n.onloadDone){n.o‌nloadDone=true;t(n.innerHTML)}}}document.getElementsByTagName("head")[0].appendChild(n)}};add_script("//github.com/dobarkod/cookie-banner/tree/master/src/cookiebanner.js");

src/cookiebanner.js
/*
Copyright (C) 2013-2015 Good Code and Cookie Banner contributors

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

(function(context) {
    'use strict';

    var win = context, doc = win.document;

    var global_instance_name = 'cbinstance';

    /*eslint-disable */
    /*!
     * contentloaded.js
     *
     * Author: Diego Perini (diego.perini at gmail.com)
     * Summary: cross-browser wrapper for DOMContentLoaded
     * Updated: 20101020
     * License: MIT
     * Version: 1.2
     *
     * URL:
     * http://javascript.nwbox.com/ContentLoaded/
     * http://javascript.nwbox.com/ContentLoaded/MIT-LICENSE
     *
     */
    // @win window reference
    // @fn function reference
    function contentLoaded(win, fn) {
        var done = false, top = true,
        doc = win.document, root = doc.documentElement,

        add = doc.addEventListener ? 'addEventListener' : 'attachEvent',
        rem = doc.addEventListener ? 'removeEventListener' : 'detachEvent',
        pre = doc.addEventListener ? '' : 'on',

        init = function(e) {
            if (e.type == 'readystatechange' && doc.readyState != 'complete') return;
            (e.type == 'load' ? win : doc)[rem](pre + e.type, init, false);
            if (!done && (done = true)) fn.call(win, e.type || e);
        },

        poll = function() {
            try { root.doScroll('left'); } catch(e) { setTimeout(poll, 50); return; }
            init('poll');
        };

        if (doc.readyState == 'complete') fn.call(win, 'lazy');
        else {
            if (doc.createEventObject && root.doScroll) {
                try { top = !win.frameElement; } catch(e) { }
                if (top) poll();
            }
            doc[add](pre + 'DOMContentLoaded', init, false);
            doc[add](pre + 'readystatechange', init, false);
            win[add](pre + 'load', init, false);
        }
    }
    /*eslint-enable */

    var Cookies = {
        get: function (key) {
            return decodeURIComponent(doc.cookie.replace(new RegExp('(?:(?:^|.*;)\\s*' + encodeURIComponent(key).replace(/[-.+*]/g, '\\$&') + '\\s*\\=\\s*([^;]*).*$)|^.*$'), '$1')) || null;
        },
        set: function (key, val, end, path, domain, secure) {
            if (!key || /^(?:expires|max-age|path|domain|secure)$/i.test(key)) {
                return false;
            }
            var expires = '';
            if (end) {
                switch (end.constructor) {
                    case Number:
                        expires = end === Infinity ? '; expires=Fri, 31 Dec 9999 23:59:59 GMT' : '; max-age=' + end;
                        break;
                    case String:
                        expires = '; expires=' + end;
                    break;
                    case Date:
                        expires = '; expires=' + end.toUTCString();
                    break;
                }
            }
            doc.cookie = encodeURIComponent(key) + '=' + encodeURIComponent(val) + expires + (domain ? '; domain=' + domain : '') + (path ? '; path=' + path : '') + (secure ? '; secure' : '');
            return true;
        },
        has: function (key) {
            return (new RegExp('(?:^|;\\s*)' + encodeURIComponent(key).replace(/[-.+*]/g, '\\$&') + '\\s*\\=')).test(doc.cookie);
        },
        remove: function (key, path, domain) {
            if (!key || !this.has(key)) { return false; }
            doc.cookie = encodeURIComponent(key) + '=; expires=Thu, 01 Jan 1970 00:00:00 GMT' + ( domain ? '; domain=' + domain : '') + ( path ? '; path=' + path : '');
            return true;
        }
    };

    var Utils = {

        // merge objects and whatnot
        merge: function() {
            var obj = {},
                i = 0,
                al = arguments.length,
                key;
            if (0 === al) {
                return obj;
            }
            for (; i < al; i++) {
                for (key in arguments[i]) {
                    if (Object.prototype.hasOwnProperty.call(arguments[i], key)) {
                        obj[key] = arguments[i][key];
                    }
                }
            }
            return obj;
        },

        str2bool: function(str) {
            str = String(str);
            switch (str.toLowerCase()) {
                case 'false':
                case 'no':
                case '0':
                case '':
                    return false;
                default:
                    return true;
                }
        },

        fade_in: function(el) {
            if (el.style.opacity < 1) {
                el.style.opacity = (parseFloat(el.style.opacity) + 0.05).toFixed(2);
                win.setTimeout(function(){
                    Utils.fade_in(el);
                }, 50);
            }
        },

        fade_out: function(el, duration_ms) {
            if (typeof(el) !== "undefined") {
                el.style.opacity = 1;
                var seconds = duration_ms / 1000;
                el.style.transition = "opacity " + seconds + "s ease";
                el.style.opacity = 0;
                setTimeout(function() { el.parentNode.removeChild(el); }, duration_ms);
            }
        },

        get_data_attribs: function(script) {
            var data = {};
            if (Object.prototype.hasOwnProperty.call(script, 'dataset')) {
                data = script.dataset;
            } else {
                var attribs = script.attributes;
                var key;
                for (key in attribs) {
                    if (Object.prototype.hasOwnProperty.call(attribs, key)) {
                        var attr = attribs[key];
                        if (/^data-/.test(attr.name)) {
                            var camelized = Utils.camelize(attr.name.substr(5));
                            // True when option name starts with "on"
                            var isFunction = (camelized.lastIndexOf('on', 0) === 0);

                            data[camelized] = isFunction ? eval(attr.value) : attr.value; // jshint ignore:line
                        }
                    }
                }
            }
            return data;
        },

        /**
        * "Standardizes" the options keys by converting and removing
        * any potential "dashed-property-name" into "dashedPropertyName".
        * In case both are present, the dashedPropertyName wins.
        */
        normalize_keys: function(options_object) {
            var camelized = {};
            for (var key in options_object) {
                if (Object.prototype.hasOwnProperty.call(options_object, key)) {
                    var camelized_key = Utils.camelize(key);
                    // TODO: could this break for "falsy" values within options_object?
                    // avoiding "dashed-property-name" overriding a potentially existing "dashedPropertyName"
                    camelized[camelized_key] = options_object[camelized_key] ? options_object[camelized_key] : options_object[key];
                }
            }
            return camelized;
        },

        camelize: function(str) {
            var separator = '-',
                match = str.indexOf(separator);
            while (match !== -1) {
                var last = (match === (str.length - 1)),
                    next = last ? '' : str[match + 1],
                    upnext = next.toUpperCase(),
                    sep_substr = last ? separator : separator + next;
                str = str.replace(sep_substr, upnext);
                match = str.indexOf(separator);
            }
            return str;
        },

        find_script_by_id: function(id) {
            var scripts = doc.getElementsByTagName('script');
            for (var i = 0, l = scripts.length; i < l; i++) {
                if (id === scripts[i].id) {
                    return scripts[i];
                }
            }
            return null;
        },

        is_numeric: function(val) {
            // eslint-disable-next-line eqeqeq
            return (parseFloat(val) == val); // want type coercion here!
        },

    };

    var script_el_invoker = Utils.find_script_by_id('cookiebanner');

    var Cookiebanner = context.Cookiebanner = function(opts) {
        this.init(opts);
    };

    Cookiebanner.prototype = {

        // for testing stuff from the outside mostly
        cookiejar: Cookies,

        init: function(opts) {

            this.inserted = false;
            this.closed = false;

            var default_text = 'We use cookies to enhance your experience. ' +
                'By continuing to visit this site you agree to our use of cookies.';
            var default_link = 'Learn more';

            this.default_options = {
                // autorun: true,
                cookie: 'cookiebanner-accepted',
                closeText: '&#10006;',
                closeStyle: 'float:right;padding-left:5px;',
                closePrecedes: true,
                cookiePath: '/',
                cookieDomain: null,
                cookieSecure: false,
                debug: false,
                expires: Infinity,
                zindex: 255,
                mask: false,
                maskOpacity: 0.5,
                maskBackground: '#000',
                height: 'auto',
                minHeight: '21px',
                bg: '#000',
                fg: '#ddd',
                link: '#aaa',
                position: 'bottom',
                padding: '5px 16px',
                message: default_text,
                linkmsg: default_link,
                moreinfo: 'http://aboutcookies.org',
                moreinfoTarget: '_blank',
                moreinfoRel: 'noopener noreferrer',
                moreinfoDecoration: 'none',
                moreinfoFontWeight: 'normal',
                moreinfoFontSize: null,
                moreinfoClass: null,
                effect: null,
                fontSize: '14px',
                fontFamily: 'arial, sans-serif',
                instance: global_instance_name,
                textAlign: 'center',
                acceptOnScroll: false,
                acceptOnClick: false,
                acceptOnTimeout: null,
                acceptOnFirstVisit: false,
                onInserted: null,
                onClosed: null,
                delayBeforeClose: null,
                fadeOutDurationMs: 2000
            };

            this.options = this.default_options;
            this.script_el = script_el_invoker;

            if (this.script_el) {
                var data_options = Utils.get_data_attribs(this.script_el);
                this.options = Utils.merge(this.options, data_options);
            }

            // allowing opts passed to the ctor to override everything
            if (opts) {
                // mimics the "data-option-name" HTML attribute becoming
                // this.options.optionName
                opts = Utils.normalize_keys(opts);
                this.options = Utils.merge(this.options, opts);
            }

            // allows customizing the global instance name via options too
            global_instance_name = this.options.instance;

            // TODO: parse/validate other options that can benefit
            this.options.zindex = parseInt(this.options.zindex, 10);
            this.options.mask = Utils.str2bool(this.options.mask);
            this.options.closePrecedes = Utils.str2bool(this.options.closePrecedes);

            // check for a possible global callback specified as a string
            if ('string' === typeof this.options.expires) {
                if ('function' === typeof context[this.options.expires]) {
                    this.options.expires = context[this.options.expires];
                } else {
                    // No global callback under that name, check if it's a
                    // number/numeric and use as duration in seconds.
                    if (Utils.is_numeric(this.options.expires)) {
                        this.options.expires = Number(parseFloat(this.options.expires));
                    }
                }
            }

            // check if expires is a callback
            if ('function' === typeof this.options.expires) {
                // TODO: this might not always be as simple as this
                this.options.expires = this.options.expires();
            }

            // Proceed with our plans only if we're invoked via a <script> element
            // that has the required id attribute.
            // For manually created instances one must call run() explicitly.
            if (this.script_el) {
                this.run();
            }
        },

        log: function() {
            /* eslint-disable no-console */
            if ('undefined' !== typeof console) {
                console.log.apply(console, arguments);
            }
            /* eslint-enable no-console */
        },

        run: function() {
            if (!this.agreed()) {
                var self = this;
                contentLoaded(win, function(){
                    self.insert();
                });
            }
        },

        build_viewport_mask: function() {
            var mask = null;
            if (true === this.options.mask) {
                var mask_opacity = this.options.maskOpacity;
                var bg = this.options.maskBackground;
                var mask_markup = '<div id="cookiebanner-mask" style="' +
                    'position:fixed;top:0;left:0;width:100%;height:100%;' +
                    'background:' + bg + ';zoom:1;filter:alpha(opacity=' +
                    (mask_opacity * 100) +');opacity:' + mask_opacity +';' +
                    'z-index:' + this.options.zindex +';"></div>';
                var el = doc.createElement('div');
                el.innerHTML = mask_markup;
                mask = el.firstChild;
            }
            return mask;
        },

        agree: function() {
            this.cookiejar.set(this.options.cookie, 1, this.options.expires, this.options.cookiePath, (this.options.cookieDomain !== '' ? this.options.cookieDomain : ''), (this.options.cookieSecure ? true : false));
            return true;
        },

        agreed: function() {
            return this.cookiejar.has(this.options.cookie);
        },

        close: function() {
            if ('fade' === this.options.effect) {
                Utils.fade_out(this.element, this.options.fadeOutDurationMs);
                this.closed = true;
            } else {
                if (this.inserted) {
                    if (!this.closed) {
                        if (this.element) {
                            this.element.parentNode.removeChild(this.element);
                        }
                        if (this.element_mask) {
                            this.element_mask.parentNode.removeChild(this.element_mask);
                        }
                        this.closed = true;
                    }
                    this.closed = true;

                    var handler = this.options.onClosed;
                    if (handler && typeof handler === 'function') {
                        handler(this);
                    }
                }
            }/* else {
                throw new Error("Not inserted but closing?!");
            }*/
            return this.closed;
        },

        agree_and_close:function() {
            if (!this.options.debug) {
                this.agree();
            }
            if (this.options.delayBeforeClose && !isNaN(parseFloat(this.options.delayBeforeClose)) && isFinite(this.options.delayBeforeClose)) {
                var self_ = this;
                setTimeout(function() { self_.close(); }, this.options.delayBeforeClose);
            } else {
                return this.close();
            }
        },

        // close and remove every trace of ourselves completely
        cleanup: function() {
            this.close();
            return this.unload();
        },

        unload: function() {
            if (this.script_el) {
                this.script_el.parentNode.removeChild(this.script_el);
            }
            context[global_instance_name] = undefined;
            // delete context[global_instance_name];
            return true;
        },

        insert: function() {
            this.element_mask = this.build_viewport_mask();

            var zidx = this.options.zindex;

            if (this.element_mask) {
                // bump notice element's zindex so it's above the mask
                zidx += 1;
            }

            var el = doc.createElement('div');
            el.className = 'cookiebanner';
            el.style.position = 'fixed';
            el.style.left = 0;
            el.style.right = 0;
            el.style.height = this.options.height;
            el.style.minHeight = this.options.minHeight;
            el.style.zIndex = zidx;
            el.style.background = this.options.bg;
            el.style.color = this.options.fg;
            el.style.lineHeight = el.style.minHeight;

            el.style.padding = this.options.padding;

            el.style.fontFamily = this.options.fontFamily;
            el.style.fontSize = this.options.fontSize;
            el.style.textAlign = this.options.textAlign;

            if ('top' === this.options.position) {
                el.style.top = 0;
            } else {
                el.style.bottom = 0;
            }

            var closeHtml = '<div class="cookiebanner-close" style="' + this.options.closeStyle + '">' +
                this.options.closeText + '</div>';

            var cssClass = '';
            if (this.options.moreinfoClass !== null) {
                cssClass = ' class="' + this.options.moreinfoClass + '" ';
            }
            var messageHtml = '<span>' + this.options.message + (this.options.linkmsg ? ' <a' + cssClass + '>' + this.options.linkmsg + '</a>' : '') + '</span>';

            if (this.options.closePrecedes) {
                el.innerHTML = closeHtml + messageHtml;
            } else {
                el.innerHTML = messageHtml + closeHtml;
            }

            this.element = el;

            var el_a;
            if (this.options.moreinfoClass !== null) {
                el_a = el.getElementsByClassName(this.options.moreinfoClass)[0];
            } else {
                el_a = el.getElementsByTagName('a')[0];
            }
            if (el_a) {
                el_a.href = this.options.moreinfo;
                el_a.target = this.options.moreinfoTarget;
                if (this.options.moreinfoRel && '' !== this.options.moreinfoRel) {
                    el_a.rel = this.options.moreinfoRel;
                }
                el_a.style.textDecoration = this.options.moreinfoDecoration;
                el_a.style.color = this.options.link;
                el_a.style.fontWeight = this.options.moreinfoFontWeight;

                if ('' !== this.options.moreinfoFontSize) {
                    el_a.style.fontSize = this.options.moreinfoFontSize;
                }
            }

            var el_x = el.getElementsByTagName('div')[0];
            el_x.style.cursor = 'pointer';

            function on(el, ev, fn) {
                var add = el.addEventListener ? 'addEventListener' : 'attachEvent',
                    pre = el.addEventListener ? '' : 'on';
                el[add](pre + ev, fn, false);
            }

            var self = this;
            on(el_x, 'click', function(){
                self.agree_and_close();
            });

            if (this.element_mask) {
                on(this.element_mask, 'click', function(){
                    self.agree_and_close();
                });
                doc.body.appendChild(this.element_mask);
            }

            // Agree and close banner on window scroll if `acceptOnScroll` option is set `true`
            if (this.options.acceptOnScroll) {
                on(window, 'scroll', function(){
                    self.agree_and_close();
                });
            }

            // Agree and close banner on click (no matter where) if `acceptOnClick` option is set `true`
            if (this.options.acceptOnClick) {
                on(window, 'click', function(){
                    self.agree_and_close();
                });
            }

            // Agree and close banner after N milliseconds
            if (this.options.acceptOnTimeout) {
                // Validate this.options.acceptOnTimeout as numeric
                if (!isNaN(parseFloat(this.options.acceptOnTimeout)) && isFinite(this.options.acceptOnTimeout)) {
                    setTimeout(function() { self.agree_and_close(); }, this.options.acceptOnTimeout);
                }
            }

            // Agree on first time the user visits a page (but do not close the cookiebanner window)
            if (this.options.acceptOnFirstVisit) {
                self.agree();
            }

            doc.body.appendChild(this.element);
            this.inserted = true;

            var handler = this.options.onInserted;
            if (handler && typeof handler === 'function') {
                handler(this);
            }

            if ('fade' === this.options.effect) {
                this.element.style.opacity = 0;
                Utils.fade_in(this.element);
            } else {
                this.element.style.opacity = 1;
            }
        }
    };

    if (script_el_invoker) {
        if (!context[global_instance_name]) {
            context[global_instance_name] = new Cookiebanner();
        }
    }

})(window);
